src/pairwise.jl

```diff
@@ -122,6 +122,32 @@ function _pairwise!(f, dest::AbstractMatrix, x, y;
     return _pairwise!(Val(skipmissing), f, dest, x′, y′, symmetric)
 end
 
+if VERSION >= v"1.6.0-DEV"
+    # Function has moved in Julia 1.7
+    if isdefined(Base, :typejoin_union_tuple)
+        using Base: typejoin_union_tuple
+    else
+        using Base.Broadcast: typejoin_union_tuple
+    end
+    # Identical to `Base.promote_typejoin` except that it uses `promote_type`
+    # instead of `typejoin`
+    function promote_type_union(::Type{T}) where T
+        if T === Union{}
+            return Union{}
+        elseif T isa UnionAll
+            return Any # TODO: compute more precise bounds
+        elseif T isa Union
+            return promote_type(promote_type_union(T.a), promote_type_union(T.b))
+        elseif T <: Tuple
+            return typejoin_union_tuple(T)
+        else
+            return T
+        end
+    end
+else
+    promote_type_union(::Type) = Any
```

Comment:
do we need to be that defensive? If passed type is not a union then I guess we can return it - right?

Reply:
Good point. Actually the only case that would be hard to support is `Tuple` types. I've refactored the code to improve this.